### ISSUE TITLE
Improve the depth model so that skip_child does not need to check for a key status

### ISSUE THREAD

**Issue by lemire** [simdjson/simdjson#1744]
Currently, we are sometimes left pointing at a key in On Demand when skip_child is called.

This requires the introduction an ugly patch:

https://github.com/simdjson/simdjson/pull/1743

There might be a better design that avoids this complexity.

**Comment by Raju1173** [comment]
Spent some time trying to understand the behavior of skip_child() and the context behind the "ugly patch".

It looks like the fundamental issue is that the current depth model lacks the information to differentiate between being at an object key vs. a string value, forcing the expensive `peek() == ':'` branch.

I also noticed the exit condition uses `depth() <= parent_depth` rather than strict equality, which seems designed to guarantee we are no longer inside the child rather than landing at a precise boundary...

Is this lack of strict positioning a side effect of this same weakness of the depth model? And looking forward, would a slightly more complex state model, one that tracks the type of token the iterator is pointing at, allow us to drop the peek() check entirely, or would tracking that extra state end up conflicting with the parser's performance goals?

**Comment by lemire** [comment]
@Raju1173 Can you elaborate about what you have in mind?

**Comment by Raju1173** [comment]
I was thinking whether the iterator model could be structured more like a small state machine, where the current position is known by construction rather than inferred.

For example, instead of handling `case "` generically and then checking whether it’s a key via peek(), the iterator could be in a state like EXPECT_KEY or EXPECT_VALUE, and the handling of case " would depend on that state. In that model, encountering a string in EXPECT_KEY state would unambiguously mean a key, and the ambiguity that leads to the peek() check wouldn’t arise in the first place.

**Comment by lemire** [comment]
> a string in EXPECT_KEY state would unambiguously mean a key, and the ambiguity that leads to the peek() check wouldn’t arise in the first place.

But you still have to check for the presence of the character `:`, right?

**Comment by Raju1173** [comment]
Yes, we would still have to verify the : to ensure the JSON is valid and to advance the pointer. We cannot avoid structural validation...

But the role of that check changes, right now, peek() == ':' is being used as a conditional branch to disambiguate our position (figuring out whether we are looking at a key or a value).

The FSM idea would not remove the need to validate ":", but it could still remove the ambiguity about whether a `"` token is a key or a value. In other words, the state would be updated on "{", ",", ":", etc., so that encountering a string in "EXPECT_KEY" would already mean it is a key, and ":" would just be a strict assrrtion to validate the structure of the file. And that check is much more predictable because most of the times the file would not have structural errors...

The real question is whether transforming that branch into a strict assertion actually saves cycles in practice. I can put together a quick proof of concept and profile the branch misses to see if this direction is actually worth exploring further.

**Comment by Raju1173** [comment]
While trying to implement this, I realized that the cost of updating a state machine every time the iterator moves just isn't worth removing a small edge case branch and it also violates the lean design of simdjson...

I tried other approaches as well and they all failed due to one of these reasons :-

1) Added bloat
2) Too much overhead

Thus, given all the directions I explored, "improving" the depth model would essentially just require shifting the peek() check somewhere else because the information required to differentiate a key from a string always depends on the next token being a colon...

I also looked into semi-branchless approaches, but the need to call `return_current_and_advance()` and log_value() upon identifying a key makes a hard branch difficult to avoid

So although the peek() check is kind of expensive, the alternatives required to eliminate that check may not be worth the cost... and the current "ugly patch" might actually be the most pragmatic balance between complexity and performance. Unless there’s a way to encode the distinction between a key and a string into the tape itself during the initial stages, the peek() might be the "least bad" option.

**Comment by lemire** [comment]
Thanks @Raju1173 for looking into this carefully. Your conclusion matches mine: telling a key from a string value requires looking at the next token (`:`), so any alternative design either moves that check elsewhere or adds per-step state tracking that costs more than it saves. We're keeping the current approach (#1743) and closing this. If someone finds a design that measurably removes the branch, a PR with benchmarks is welcome.
